uber/jinja.py

```diff
@@ -1,4 +1,106 @@
 from uber.common import *
+import traceback
+import weakref
+from jinja2.loaders import split_template_path
+from jinja2.utils import open_if_exists
+
+
+class MultiPathEnvironment(jinja2.Environment):
+
+    def _find_requesting_templates(self):
+        """
+        Returns the set of all files in the call stack ending in ".html".
+        """
+        stack = map(lambda s: s[0], traceback.extract_stack())
+        return set(filter(lambda s: s.endswith('.html'), stack))
+
+    def _load_template(self, name, globals):
+        """
+        Overridden to take into consideration the templates in the call stack.
+        """
+        if self.loader is None:
+            raise TypeError('no loader for this environment specified')
+
+        requesting_templates = self._find_requesting_templates()
+
+        cache_key = (weakref.ref(self.loader), name, ','.join(requesting_templates))
+        if self.cache is not None:
+            template = self.cache.get(cache_key)
+            if template is not None and (not self.auto_reload or
+                                         template.is_up_to_date):
+                return template
+        template = self.loader.load(self, name, globals, requesting_templates)
+        if self.cache is not None:
+            self.cache[cache_key] = template
+        return template
+
+
+class MultiPathLoader(jinja2.FileSystemLoader):
+
+    def get_source(self, environment, template, requesting_templates=set()):
+        """
+        Overridden to accept the requesting_templates parameter: a set of every
+        template that has already been loaded during the current request.
+        """
+        pieces = split_template_path(template)
+        for searchpath in self.searchpath:
+            filename = os.path.join(searchpath, *pieces)
+            if filename in requesting_templates:
+                # If the file is already in the call stack, ignore it
+                continue
+
+            f = open_if_exists(filename)
+            if f is None:
+                continue
+            try:
+                contents = f.read().decode(self.encoding)
+            finally:
```

Comment:
Should there be an `except` block here? What happens if there's a problem loading a template? Is that passed up the stack correctly so we see the exception rendered?

Reply:
That exception would be passed up the stack, so we'd see the exception rendered. Most of the code in `MultiPathEnvironment` and `MultiPathLoader` was lifted wholesale from the Jinja source code

See: 
* https://github.com/pallets/jinja/blob/master/jinja2/loaders.py#L167
* https://github.com/pallets/jinja/blob/master/jinja2/environment.py#L795